lib/puppet/provider/cisco_itd_service/cisco.rb

```diff
@@ -0,0 +1,256 @@
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@nu',
+                                            ITDSERVICE_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:array_flat, self, '@nu',
+                                            ITDSERVICE_ARRAY_FLAT_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@nu',
+                                            ITDSERVICE_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@nu',
+                                            ITDSERVICE_SHUT_PROP)
+
+  def initialize(value={})
+    super(value)
+    @nu = Cisco::ItdService.itds[@property_hash[:name]]
+    @property_flush = {}
+  end
+
+  def self.properties_get(itd_service_name, nu_obj)
+    debug "Checking instance, #{itd_service_name}."
+    current_state = {
+      service_name: itd_service_name,
+      name:         itd_service_name,
+      ensure:       :present,
+    }
+    # Call node_utils getter for each property
+    ITDSERVICE_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+    ITDSERVICE_ARRAY_FLAT_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+    ITDSERVICE_ALL_BOOL_PROPS.each do |prop|
+      val = nu_obj.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    # nested array properties
+    current_state[:ingress_interface] = nu_obj.ingress_interface
+    current_state[:peer_vdc] = nu_obj.peer_vdc
+    current_state[:virtual_ip] = nu_obj.virtual_ip
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    itds = []
+    Cisco::ItdService.itds.each do |itd_service_name, nu_obj|
+      itds << properties_get(itd_service_name, nu_obj)
+    end
+    itds
+  end
+
+  def self.prefetch(resources)
+    itds = instances
+    resources.keys.each do |name|
+      provider = itds.find { |nu_obj| nu_obj.instance_name == name }
+      resources[name].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def exists?
+    (@property_hash[:ensure] == :present)
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def instance_name
+    service_name
+  end
+
+  def shutdown_set
+    cur_shut = @property_hash[:shutdown]
+    next_shut = @resource[:shutdown]
+    cur_shut == :false && (next_shut.nil? || next_shut == :true)
+  end
+
+  def all_prop_set(new_itd)
+    ITDSERVICE_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      send("#{prop}=", @resource[prop]) if new_itd
+      unless @property_flush[prop].nil?
+        @nu.send("#{prop}=", @property_flush[prop]) if
+          @nu.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    load_balance_set
+  end
+
+  def shut_prop_set(new_itd)
```

Comment:
With the change above no need to loop through an array that only contains a single property:
```
def shutdown_set(new_itd)
  return unless @resource[:shutdown]
  send('shutdown=', @resource[:shutdown]) if new_itd
  unless @property_flush[:shutdown].nil?
    @nu.send('shutdown=', @property_flush[:shutdown]) if
      @nu.respond_to?('shutdown=')
  end
end
```